lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py

```diff
@@ -130,25 +192,99 @@ def get_latest_installable_revision( self, trans, payload, **kwd ):
                 return changeset_revisions[ -1 ]
         return hg_util.INITIAL_CHANGELOG_HASH
 
-    def __flatten_repository_dependency_list( self, trans, tool_shed_repository ):
-        '''
-        Return a recursive exclusive flattened list of all tool_shed_repository's dependencies.
-        '''
-        dependencies = []
-        for dependency in tool_shed_repository.repository_dependencies:
-            if len( dependency.repository_dependencies ) > 0:
-                sub_dependencies = self.__flatten_repository_dependency_list( trans, dependency )
-                for sub_dependency in sub_dependencies:
-                    if sub_dependency not in dependencies:
-                        dependencies.append( sub_dependency )
-            if dependency not in dependencies:
-                dependencies.append( dependency.as_dict( value_mapper=self.__get_value_mapper( trans, tool_shed_repository ) ) )
-        return dependencies
+    @expose_api
+    @web.require_admin
+    def shed_categories( self, trans, **kwd ):
+        """
+        GET /api/tool_shed_repositories/shed_categories
 
-    def __get_value_mapper( self, trans, tool_shed_repository ):
-        value_mapper = { 'id': trans.security.encode_id( tool_shed_repository.id ),
-                         'error_message': tool_shed_repository.error_message or '' }
-        return value_mapper
+        Display a list of categories in the selected toolshed.
+
+        :param tool_shed_url: the url of the toolshed to get categories from
+        """
+        tool_shed_url = kwd.get( 'tool_shed_url', '' )
+        tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( trans.app, tool_shed_url )
+        url = util.build_url( tool_shed_url, pathspec=[ 'api', 'categories' ] )
+        categories = json.loads( util.url_get( url ) )
+        repositories = []
+        url = util.build_url( tool_shed_url, pathspec=[ 'api', 'repositories' ] )
+        for repo in json.loads( util.url_get( url ) ):
+            repositories.append( dict( value=repo[ 'id' ], label='%s/%s' % ( repo[ 'owner' ], repo[ 'name' ] ) ) )
+        return { 'categories': categories, 'repositories': repositories }
+
+    @expose_api
+    @web.require_admin
+    def shed_category( self, trans, **kwd ):
```

Comment:
My thinking was that this is an endpoint (currently) used exclusively for repository installation, which is itself an admin-only function. Thus it made sense to make this endpoint also admin-only.